src/Hamiltonians/Molecular.jl

```diff
@@ -0,0 +1,613 @@
+)::T where {T<:Number}
+    two_elec_int = zero(T)
+
+    sum_alpha_alpha = zero(T)
+    for i in occ_modes[1]
+        for j in occ_modes[1]
+            if i.mode ≠ j.mode
+                sum_alpha_alpha += int2[i.mode, j.mode, i.mode, j.mode] -
+                                   int2[i.mode, j.mode, j.mode, i.mode]
+            end
+        end
+    end
+    two_elec_int += 0.5 * sum_alpha_alpha
+
+    sum_alpha_beta = zero(T)
+    for i in occ_modes[1]
+        for j in occ_modes[2]
+            sum_alpha_beta += int2[i.mode, j.mode, i.mode, j.mode]
+        end
+    end
+    two_elec_int += sum_alpha_beta
+
+    sum_beta_beta = zero(T)
+    for i in occ_modes[2]
+        for j in occ_modes[2]
+            if i.mode ≠ j.mode
+                sum_beta_beta += int2[i.mode, j.mode, i.mode, j.mode] -
+                                 int2[i.mode, j.mode, j.mode, i.mode]
+            end
+        end
+    end
+    two_elec_int += 0.5 * sum_beta_beta
+
+    two_elec_int
+end
+
+@inline function flip_spin_chan(chan::Int)::Int
+    if chan == 1
+        return 2
+    elseif chan == 2
+        return 1
+    end
+end
+
+"""
+This struct is used internally to represent the iterator for 
+off-diagonal terms generation.
+"""
+struct MolecularHamiltonianOffDiagonalsIterator{
+    T,A<:FermiFS2C,D,OA,OB,UA,UB,TI<:FermiFSIndex
+}
+    address::A
+    op::MolecularHamiltonian{T,A,D}
+    modes::Modes{OA,OB,UA,UB,TI}
+end
+
+function MolecularHamiltonianOffDiagonalsIterator(
+    addr::A, op::MolecularHamiltonian{T,A,D}, m::Modes{OA,OB,UA,UB,TI}
+) where {T,A,D,OA,OB,UA,UB,TI}
+    return MolecularHamiltonianOffDiagonalsIterator{T,A,D,OA,OB,UA,UB,TI}(addr, op, m)
+end
+
+function MolecularHamiltonianOffDiagonalsIterator(
+    c::MolecularHamiltonianOperatorColumn
+)
+    return MolecularHamiltonianOffDiagonalsIterator(c.address, c.op, c.modes)
+end
+
+function Base.eltype(
+    ::MolecularHamiltonianOffDiagonalsIterator{T,A}
+) where {T,A}
+    return Tuple{A,T}
+end
+
+"""
+This struct is used internally to represent the state during 
+off-diagonal terms generation.
+
+* `n_excited`: Tuple of 2 indices represents the how many electrons are excited in 
+alpha(1) and beta(2) channel.
+* `from`: Records the indices of `ModeMap` arrays where electrons are excited from.
+* `to`: Records the indices of `ModeMap` arrays where electrons are excited to.
+
+Both `from = (0,0)` and `to = (0,0)` represent a special state when there is no more
+state in current `n_excited` situation, this is used to notify upper caller.
+
+When used to represents one-electron-excitation case, only `from[1]` and `to[1]`
+are used. 
+When used to represents one-one-electrons-excitation case, index `1` represention
+alpha spin channel and index `2` represention beta spin channel repectively.
+"""
+struct MolecularHamiltonianOffDiagonalsIteratorState
+    n_excited::Tuple{Int,Int}
+    from::Tuple{Int,Int}
+    to::Tuple{Int,Int}
+end
```

Comment:
I suggest renaming for clarity, as "self-documenting code" (check if correct!):
```suggestion
struct MolecularHamiltonianOffDiagonalsIteratorState
    excitations_per_channel::Tuple{Int,Int} # either 0, 1, 2
    from_occupieds::Tuple{Int,Int} # indices into ModeMap
    to_unoccupieds::Tuple{Int,Int} # indices into ModeMap
end
```